src/ArrayView.hpp

```diff
@@ -296,28 +295,45 @@ class ArrayView
   /**
    * @brief Overload for rvalues that raises a compilation error when used.
    * @return A null ArraySlice.
-   * @brief @c toSliceConst cannot be called on a rvalue since the @c ArraySlice would
-   *   contain pointers to the object that is about to be destroyed. This overload
-   *   prevents that from happening.
+   * @brief This cannot be called on a rvalue since the @c ArraySlice would
+   *   contain pointers to the dims and strides of the current @c ArrayView that is
+   *   about to be destroyed. This overload prevents that from happening.
    */
-  template< typename _T=T >
   inline LVARRAY_HOST_DEVICE constexpr
   ArraySlice< T const, NDIM, USD, INDEX_TYPE >
   toSliceConst() const && noexcept
   {
-    static_assert( NDIM < 0, "Cannot call toSliceConst on a rvalue." );
+    static_assert( !typeManipulation::always_true< T >, "Cannot call toSliceConst on a rvalue." );
     return ArraySlice< T const, NDIM, USD, INDEX_TYPE >( nullptr, nullptr, nullptr );
   }
 
   /**
-   * @return Return *this interpret as ArrayView<T const> const &.
+   * @brief A user defined conversion operator (UDC) to an ArrayView< T const, ... >.
+   * @return A new ArrayView where @c T is @c const.
    */
   template< typename _T=T >
   inline LVARRAY_HOST_DEVICE constexpr
   operator std::enable_if_t< !std::is_const< _T >::value,
-                             ArrayView< T const, NDIM, USD, INDEX_TYPE, BUFFER_TYPE > >() const noexcept
+                             ArrayView< T const, NDIM, USD, INDEX_TYPE, BUFFER_TYPE > >() const & noexcept
   { return toViewConst(); }
 
+  /**
+   * @brief Overload for rvalues that raises a compilation error when used.
+   * @return A null ArrayView.
+   * @brief This is valid when this is an @c ArrayView but invalid when this is an @c Array.
+   *   I have been unable to allow the @c ArrayView conversion while not allowing the @c Array.
+   *   Therefore the @c ArrayView conversion is not allowed, instead simply call @c toViewConst.
+   */
+  template< typename _T=T >
+  inline LVARRAY_HOST_DEVICE constexpr
+  operator std::enable_if_t< !std::is_const< _T >::value,
```

Comment:
This is valid for an `ArrayView` but not an `Array`. I wasn't able to find a solution that allowed this with out errors in either Clang or GCC and warnings in NVCC. I chose to disallow it for `ArrayView` since you can always explicitly call `toViewConst`. This is the source of most of the changes in GEOSX.